src/jdk.compiler/share/classes/com/sun/tools/javac/code/DeferredLintHandler.java

```diff
@@ -52,99 +71,107 @@ public static DeferredLintHandler instance(Context context) {
         return instance;
     }
 
-    /** The Lint to use when {@link #immediate(Lint)} is used,
-     * instead of {@link #setPos(DiagnosticPosition)}. */
-    private Lint immediateLint;
+    /**
+     * Registered {@link LintLogger}s grouped by the innermost containing module, package, class,
+     * method, or variable declaration.
+     */
+    private final HashMap<JCTree, ArrayList<LintLogger>> deferralMap = new HashMap<>();
+
+    /**
+     * The current "reporter" stack, reflecting calls to {@link #push} and {@link #pop}.
+     *
+     * <p>
+     * The top of the stack determines how calls to {@link #report} are handled.
+     */
+    private final ArrayDeque<Consumer<LintLogger>> reporterStack = new ArrayDeque<>();
 
     @SuppressWarnings("this-escape")
     protected DeferredLintHandler(Context context) {
         context.put(deferredLintHandlerKey, this);
-        this.currentPos = IMMEDIATE_POSITION;
-        immediateLint = Lint.instance(context);
+        Lint rootLint = Lint.instance(context);
+        pushImmediate(rootLint);            // default to "immediate" mode
     }
 
+// LintLogger
+
     /**An interface for deferred lint reporting - loggers passed to
      * {@link #report(LintLogger) } will be called when
      * {@link #flush(DiagnosticPosition) } is invoked.
      */
     public interface LintLogger {
+
+        /**
+         * Generate a warning if appropriate.
+         *
+         * @param lint the applicable lint configuration
+         */
         void report(Lint lint);
     }
 
-    private DiagnosticPosition currentPos;
-    private Map<DiagnosticPosition, ListBuffer<LintLogger>> loggersQueue = new HashMap<>();
+// Reporter Stack
 
-    /**Associate the given logger with the current position as set by {@link #setPos(DiagnosticPosition) }.
-     * Will be invoked when {@link #flush(DiagnosticPosition) } will be invoked with the same position.
-     * <br>
-     * Will invoke the logger synchronously if {@link #immediate() } was called
-     * instead of {@link #setPos(DiagnosticPosition) }.
+    /**
+     * Defer {@link #report}ed warnings until the given declaration is flushed.
+     *
+     * @param decl module, package, class, method, or variable declaration
+     * @see #pop
      */
-    public void report(LintLogger logger) {
-        if (currentPos == IMMEDIATE_POSITION) {
-            logger.report(immediateLint);
-        } else {
-            ListBuffer<LintLogger> loggers = loggersQueue.get(currentPos);
-            if (loggers == null) {
-                loggersQueue.put(currentPos, loggers = new ListBuffer<>());
-            }
-            loggers.append(logger);
-        }
+    public void push(JCTree decl) {
```

Comment:
I wonder if `enterDecl`/`leaveDecl` would be more evocative than `push`/`pop`. (but don't have any suggestion for the weird immediate mode).